cni/pkg/ambient/informers.go

```diff
@@ -120,31 +121,49 @@ func (s *Server) Reconcile(input any) error {
 			s.AddPodToMesh(pod)
 		}
 	case controllers.EventUpdate:
-		// For update, we just need to handle opt outs
-		newPod := event.New.(*corev1.Pod)
-		oldPod := event.Old.(*corev1.Pod)
-		ns := s.namespaces.Get(newPod.Namespace, "")
-		if ns == nil {
-			return fmt.Errorf("failed to find namespace %v", ns)
-		}
-		wasEnabled := oldPod.Annotations[constants.AmbientRedirection] == constants.AmbientRedirectionEnabled
-		nowEnabled := PodRedirectionEnabled(ns, newPod)
-		if wasEnabled && !nowEnabled {
-			log.Debugf("Pod no longer matches, removing from mesh")
-			s.DelPodFromMesh(newPod, event)
-		} else if !wasEnabled && nowEnabled {
-			log.Debugf("Pod now matches, adding to mesh")
-			s.AddPodToMesh(pod)
-		} else if nowEnabled && !IsPodInIpset(pod) {
-			// This can happen if a node cleanup happens as part of a ztunnel recycle,
-			// cleaning up the node-level ipset with all the pod IPs
-			// If this happens we re-queue everything for reconciliation, and need to
-			// make sure existing pods get re-added to the ipset if they aren't already there.
-			log.Debugf("Pod is enabled but not in ipset, (re)adding to mesh")
-			s.AddPodToMesh(pod)
-		}
+		return s.handleUpdate(event)
 	case controllers.EventDelete:
 		s.DelPodFromMesh(pod, event)
 	}
 	return nil
 }
+
+func (s *Server) handleUpdate(event controllers.Event) error {
```

Comment:
If we are going to split this logic out into a smaller function when it only has one caller, we should justify the change by unit testing it.

Reply:
1. `"istio.io/istio/pkg/kube"` already  has a `NewFakeClient` which can be used to generate pod change events, it is used in other Istio tests.

2. splitting `setupHandlers` and `go s.queue.Run()` out from the rest and passing them the faked KubeClient from 1 should let you generate pod change events and validate.

3. As far as mocking I'm fond of `testify/mock` over DIY mocking.

Reply:
Thank you so much @bleggett 
I added the test case